src/Symfony/Component/Messenger/Envelope.php

```diff
@@ -72,4 +85,13 @@ public function getMessage()
     {
         return $this->message;
     }
+
+    public function getDispatchKey(): ?string
+    {
+        if (null !== $key = $this->stamps[DispatchKeyStamp::class] ?? null) {
+            $key = $key->getName();
+        }
+
+        return $key;
```

Comment:
That would break SenderLocator/HandlerLocator, which need to know whether the key has been set explicitly or if it should fallback to type-based dispatching.

Reply:
i was thinking

```php
$name = $envelope->getDispatchKey();
if (isset($mapping[$name])) {
  return $mapping[$name];
}

$class = \get_class($envelope->getMessage());

foreach (class_parents($class) as $name) {
  if (isset($mapping[$name])) {
    return $mapping[$name];
  }
}

// ...

return $mapping['*'] ?? null;
```

Reply:
yes but no :)
I mean: we'd better kill type-based dispatching than make a lying getDispatchKey method that would return something that is not what is used for dispatching...

Reply:
I agree with you it'd be better, but type-based dispatching means we can't "always use getDispatchKey()" for dispatching. Note I'm serious about dropping reflection-based dispatching.
Until it happens (either we agree before 4.2 or via later deprecation), this makes more sense as is to me.

Reply:
i still dont get it :( per phpdoc:

> if not provided, the key is derived from the class of the message

a nullable `getDispatchKey()` contradicts with this IMHO

That's unrelated to type-based dispatching based on reflection no?

Reply:
> the key is derived from the class of the message

this doesn't meand actually you can get it: with reflection-based dispatching, the key is discoevered at runtime, as the intersection of types implemented by the message and what handlers/senders are listening to. So: "key" in this sentence has a broader meaning that the getDispatchKey method alone cannot resolve.

Reply:
what confuses me, is that _after_ dispatch the key (obtained from the envelop) can still be null and we never get the "resolved" key per getDispatchKey

Reply:
Correct, that's because of the dynamic dispatching that happens right now with reflection-based dispatching. See my previous comments. We could do what you suggest but that would just cloud the issue.